transcom/mymove
# INT B-21162 Destination Address Street 1 OPTIONAL

## [B-21162](https://www13.v1host.com/USTRANSCOM38/story.mvc/summary?oidToken=Story%3A1011460)

## Summary

PPM destination address street 1 will now be OPTIONAL for onboarding on the CUSTOMER, OFFICE, PRIME SIMULATOR APPS and PRIME V3 APIs(CreateMTOShipment a d UpdateMTOShipment). **PPM destination address street 1 will be REQUIRED during PRE CLOSEOUT(waiting on customer) on the CUSTOMER side during the Upload PPM Documents process.**

It was decided not to make streetAddress1 NULLABLE on the addresses table per architecture's _suggestion_.  Instead  we will be using a string placeholder - "n/a" - to represent NULL on the table level. To achieve this,  the payload_to_model and model_to_payload GO files were updated to convert empty/null value to "n/a and vice versa. 
It will be shown as streetAddress1 = "" on the client side response side. This was done per Danny M not wanting the UI to display "n/a".

Note: This PR takes in consideration of new enchancement introduced by the ByteSize MM's [MAIN PR](https://github.com/transcom/mymove/pull/13756). This enhancement will now display REQUIRED labelHints for all required input fields while removing the OPTIONAL label hints; However, this change was only done on the CUSTOMER SIDE and NOT on OFFICE/PRIME SIMULATOR UIs.

**CUSTOMER/ONBOARDING:**
![customer_ppm](https://github.com/user-attachments/assets/4202ae97-9b22-432e-9c93-8e9ca854510f)
**CUSTOMER/UPLOAD PPM DOCUMENTS:**
![customer_upload_docs_ppm](https://github.com/user-attachments/assets/ad2505a5-0e97-4821-b36b-e83dc72ce978)


**OFFICE/PRIME SIMULATOR:**
![office_ppm](https://github.com/user-attachments/assets/8c36bda5-9bd2-46a1-994f-37f53aadce53)



[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

1. Login as customer and onboard PPM. Verify PPM can be saved with or without destination address street 1.   **Required** labelHints must be displayed for all required input controls. See attached screenshot above.
2. Submit PPM and login as service counselor. Create/update PPM and verify PPM can be saved with or without providing destination address street 1.  Verify OPTIONAL labelHints are displayed for all OPTIONAL input controls. See screenshot above.
3. Login as Prime Simulator UI and create/update a PPM. Like the CUSTOMER and OFFICE, verify PPM can be saved with and without destination address street 1. Check if OPTIONAL labelHints are displayed for all OPTIONAL input controls. See 
4. Submit Move Details
5. Login as customer and verify destination address street 1 is REQUIRED in the Upload PPM  Documents process.
see screenshot above.

How to test Prime V3 API:
1. Login to Prime Simulator UI. Pick an existing move and goto details view. Copy moveId from UI.
2. Goto http://primelocal:3000/swagger-ui/prime_v3.html#/mtoShipment/createMTOShipment . Paste Move ID from step 1 for moveTaskOrderID field in sample JSON and replace "<-- PASTE HERE-->":
`{
  "moveTaskOrderID": "<-- PASTE HERE-->",
  "shipmentType": "PPM",
  "counselorRemarks": null,
  "ppmShipment": {
    "expectedDepartureDate": "2024-10-01",
    "pickupAddress": {
      "streetAddress1": "123 Street",
      "streetAddress2": "",
      "city": "SomeCity",
      "state": "CA",
      "postalCode": "90210"
    },
    "secondaryPickupAddress": null,
    "destinationAddress": {
      "streetAddress1": "",
      "streetAddress2": "",
      "city": "SomeCity",
      "state": "CA",
      "postalCode": "90210"
    },
    "secondaryDestinationAddress": null,
    "sitExpected": false,
    "hasSecondaryPickupAddress": false,
    "hasSecondaryDestinationAddress": false,
    "estimatedWeight": 100,
    "hasProGear": false
  }
}`
3. click Execute. Verify PPM was created via Prime Simulator UI. Copy shipment ID and Shipment eTag for newly created PPM.
4. Goto http://primelocal:3000/swagger-ui/prime_v3.html#/mtoShipment/updateMTOShipment . Paste shipment ID and Shipment eTag from step 3.  Use your own JSON or the following:
`{
  "ppmShipment": {
    "expectedDepartureDate": "2024-10-01",
    "pickupAddress": {
      "streetAddress1": "123 Street",
      "city": "SomeCity",
      "state": "AL",
      "postalCode": "14150"
    },
    "secondaryPickupAddress": {
      "streetAddress1": "",
      "streetAddress2": "",
      "streetAddress3": "",
      "city": "",
      "state": "",
      "postalCode": ""
    },
    "tertiaryPickupAddress": {
      "streetAddress1": "",
      "streetAddress2": "",
      "streetAddress3": "",
      "city": "",
      "state": "",
      "postalCode": ""
    },
    "destinationAddress": {
      "streetAddress1": "",
      "city": "SomeCity test from swagger 2b",
      "state": "AL",
      "postalCode": "14150"
    },
    "secondaryDestinationAddress": {
      "streetAddress1": "",
      "streetAddress2": "",
      "streetAddress3": "",
      "city": "",
      "state": "",
      "postalCode": ""
    },
    "tertiaryDestinationAddress": {
      "streetAddress1": "",
      "streetAddress2": "",
      "streetAddress3": "",
      "city": "",
      "state": "",
      "postalCode": ""
    },
    "sitExpected": false,
    "estimatedWeight": 100,
    "hasProGear": false,
    "hasSecondaryPickupAddress": false,
    "hasTertiaryPickupAddress": false,
    "hasSecondaryDestinationAddress": false,
    "hasTertiaryDestinationAddress": false
  },
  "counselorRemarks": null
}
`
6. click Execute. Refresh Prime Simulator UI and verify expected changes are present.
7. Try to break it. Test with empty string, white spaces and without streetAddress1 property. 


### Frontend


### Backend
